---
title: "Grun-mESC-2014"
author: "Belinda Phipson"
date: "08/02/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
The data is mouse embryonic stem cells with UMIs from Gr√ºn D, Kester L, van Oudenaarden A. Validation of noise models for single-cell transcriptomics. Nat Methods 2014 Jun;11(6):637-40. PMID: 24747814

I downloaded the counts from GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE54695
The counts are not the raw UMI counts. The authors estimated transcript number from the UMI counts based on binomial statistics.

The reads were mapped to the mouse genome (mm10) using BWA with default parameters.

There are two groups of cells, each subjected to one of two growth media: 2i and serum.

# Load R libraries, and useful R functions

```{r,libraries,cache=FALSE}
library(edgeR)
library(RColorBrewer)
library(org.Mm.eg.db)
library(limma)
```
```{r,myRfunctions}
source("/group/bioi1/belinda/SingleCell/Normalisation/normCounts.R")
```
```{r}
ggplotColors <- function(g){

  d <- 360/g

  h <- cumsum(c(15, rep(d,g - 1)))

  hcl(h = h, c = 100, l = 65)

}
```
# Reading in the data and sample information
```{r}
counts <- read.delim("/group/bioi1/belinda/SingleCell/GeneLength/GSE54695_data_transcript_counts.txt",row.names=1)

sc <- grep("SC",colnames(counts))
rna <- grep("RNA",colnames(counts))

sc.counts <- counts[,sc]
rna.counts <- counts[,rna]

# load mm10 gene length information
mm10.length <- read.table("mm10-length.txt",header=TRUE,stringsAsFactors = FALSE)
```
```{r}
# Separating endogenous and ERCC controls
ercc <- grep("ERCC",rownames(sc.counts))
counts.ercc <- sc.counts[ercc,]
counts.end <- sc.counts[-ercc,]
```

# Quality control of the cells
## QC plots

Before any analysis, the quality of the cells was examined by looking at metrics such as dropout and total library size of each cell. We also looked at the proportion of the sequencing taken up by the ERCC controls. 

```{r}
# Calculate dropout and library size
dropout <- colSums(counts.end==0)/nrow(counts.end)
lib.size <- colSums(sc.counts)
lib.size.ercc <- colSums(counts.ercc)
lib.size.end <- colSums(counts.end)
```

```{r}
par(mar=c(5,4,2,2))
par(mfrow=c(1,2))
plot(dropout,lib.size.end,xlab="Proportion Zeroes in Single Cells",ylab="Library size")
abline(h=10000,v=0.8,lty=2)
plot(dropout,lib.size.ercc/lib.size,ylab="ERCC lib size/Total lib size",xlab="Proportion Zeroes in Single Cells")
abline(v=0.8,h=0.05,lty=2)
```

## Filtering out cells
We decide to filter out cells with > 80% dropout, smaller than 10000 library size, as cells that had a ratio of ERCC library size to total library size that had > 0.05.
```{r}
keep1 <- dropout<0.80 & lib.size.end>10000 & lib.size.ercc/lib.size < 0.05
counts.keep <- counts.end[,keep1]
dim(counts.keep)
```
This leaves `r ncol(counts.keep)` cells for further analysis.

## Filtering out genes
We chose to filter out genes that had more than 90% zeroes.
```{r}
# Filter out genes where > 90% of the cells are zeroes
propZ_genes <- rowSums(counts.keep==0)/ncol(counts.keep)
counts.keep <- counts.keep[propZ_genes<=0.9,]
dim(counts.keep)
```
This left `r nrow(counts.keep)` genes for further analysis. We finally check an MDSplot with PC1 and PC2.
```{r}
par(mfrow=c(1,1))
media <- rep(NA,ncol(counts.keep))
media[grep("2i",colnames(counts.keep))] <- c("2i")
media[grep("serum",colnames(counts.keep))] <- c("serum")
plotMDS(DGEList(counts.keep),pch=16,gene.selection = "common",col=ggplotColors(2)[factor(media)])
legend("bottom",legend=levels(factor(media)),fill=ggplotColors(2))
```

# Converting to DGEList, adding gene annotation
```{r,cache=TRUE}
y <- DGEList(counts.keep,group=media)
```

```{r,cache=TRUE}
#annotation
symbol <- toTable(org.Mm.egSYMBOL)
m <- match(rownames(y),symbol$symbol)
ann <- data.frame(Original_ID=rownames(y),symbol[m,])
rownames(ann) <- rownames(y)
ens <- toTable(org.Mm.egENSEMBL)
m <- match(ann$gene_id,ens$gene_id)
ann$ensembl_id <- ens$ensembl_id[m]
chr <- toTable(org.Mm.egCHR)
m <- match(ann$gene_id,chr$gene_id)
ann$chr <- chr$chromosome[m]
genename <- toTable(org.Mm.egGENENAME)
m <- match(ann$gene_id,genename$gene_id)
ann$genename <- genename$gene_name[m]
m <- match(ann$ensembl_id,mm10.length$EnsID)
ann$length <- mm10.length$Length[m]

y$genes <- ann
```

# Filtering out additional genes based on annotation
We decided to filter out mitochondrial and ribosomal genes, as well as genes that lack an Entrez gene id. This was to clean up the dataset as much as possible and stick to well annotated genes.

```{r}
mito <- grep("mitochondrial",y$genes$genename)
ribo <- grep("ribosomal",y$genes$genename)
chrm <- grep("MT",y$genes$chr)
junk <- unique(c(mito,ribo,chrm))
length(junk)
y <- y[-junk,]
y <- y[!is.na(y$genes$gene_id),]
y <- y[!is.na(y$genes$length),]
dim(y)
y$samples$lib.size <- colSums(y$counts)
y.grun <- y
save(y.grun,file="./mESCs-data/grun.Rdata")
```

# Gene length bias
We binned genes into 10 roughly equal bins based on gene length, with roughly 1000 genes in each bin.
```{r,cache=TRUE}
lcpm <- normCounts(y,log=TRUE,prior.count=1)
lrpkm <- rpkm(y,log=TRUE,prior.count=1)
propZ_genes <- rowSums(y$counts==0)/ncol(y)
```

```{r}
# Binning according to gene length
q<-quantile(sqrt(y$genes$length),probs=seq(0.1,1,0.1))
decile <- rep(NA,nrow(y))
decile[sqrt(y$genes$length)<=q[1]] <- 1
for(i in 2:10) decile[sqrt(y$genes$length)>q[i-1] & sqrt(y$genes$length)<=q[i]] <- i
```

# Figure for paper:

```{r,fig1A,cache=FALSE,fig.width=12,fig.height=5}
par(mar=c(8.5,4.5,3,2))
par(mfrow=c(1,3))
par(mgp=c(3,1,0))

q2 <- c(0,q^2)
labels <- rep(NA,10)
for(i in 1:10) labels[i] <-paste(round(q2[i]),"-",round(q2[i+1]),sep="")

boxplot(rowMeans(lcpm)~decile,ylab="AvgLogCounts",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(0,5))
title("(a) Mouse ESCs, umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(propZ_genes~decile,ylab="Proportion zeroes in each gene",cex.lab=1.5,cex.axis=1.2,names=labels,las=2)
title("(b) Mouse ESCs, umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(rowMeans(lrpkm)~decile,ylab="AvgLogRPKM",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(-1,10))
title("(c) Mouse ESCs, umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
```

```{r}
save(lcpm,decile,labels,propZ_genes,lrpkm,file="./figuresForPaper/Grun-fig2Ai.Rdata")
```








