---
title: "Klein-hK562-2015"
author: "Belinda Phipson"
date: "07/02/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

The data is from : Klein AM et al. 2015. Droplet barcoding for single-cell transcriptomics applied to embryonic stem cells. Cell. 161:1187–1201. http://dx.doi.org/10.1016/j.cell.2015.04.044. 

I downloaded the counts table (unnormalised UMI counts) from GEO under accession GSM1599500: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1599500. The data was aligned to the human genome (hg19) using Bowtie.

The library preparation is described as follows: Cells were encapsulated into droplets on ice and lysed in the 4nL microfluidic droplets using a final concentration of 0.4% NP-40. Single cell lysates were subject to reverse transcription at 50°C without purification of RNA.
Cells were barcoded using the db-Seq platform, which makes use of the CEL-Seq protocol for library construction (Hashimshony et al., Cell Reports 2011).

# Load R libraries, and useful R functions

```{r,cache=FALSE}
library(edgeR)
library(RColorBrewer)
library(org.Hs.eg.db)
library(limma)
```
```{r}
source("/group/bioi1/belinda/SingleCell/Normalisation/normCounts.R")
```
```{r}
ggplotColors <- function(g){

  d <- 360/g

  h <- cumsum(c(15, rep(d,g - 1)))

  hcl(h = h, c = 100, l = 65)

}
```
# Reading in the data
```{r,cache=TRUE}
counts.klein <- read.csv("/group/bioi1/belinda/SingleCell/Klein-UMI-scRNAseq/GSM1599500_K562_cells.csv",stringsAsFactors = FALSE,row.names=1)
colnames(counts.klein) <- paste("Cell",1:ncol(counts.klein),sep="")

# Load in featureCounts hg38 length information
hg38.length <- read.table("hg38-length.txt",header=TRUE,stringsAsFactors = FALSE)
```

```{r,cache=TRUE}
# Separate ercc and endogenous genes
ercc <- grep("ERCC", rownames(counts.klein))
counts.end <- counts.klein[-ercc,]
counts.ercc <- counts.klein[ercc,]
```

# Quality control
## QC plots
Before any analysis, the quality of the cells was examined by looking at metrics such as dropout and total library size of each cell. We also looked at the proportion of the sequencing taken up by the ERCC controls.

```{r}
# Calculate dropout and library size
dropout <- colSums(counts.end==0)/nrow(counts.end)
lib.size <- colSums(counts.klein)
lib.size.ercc <- colSums(counts.ercc)
lib.size.end <- colSums(counts.end)
```

```{r}
par(mar=c(5,4,2,2))
par(mfrow=c(1,2))
plot(dropout,lib.size.end,xlab="Proportion Zeroes in Single Cells",ylab="Library size")
abline(h=10000,v=0.85,lty=2)
plot(dropout,lib.size.ercc/lib.size,ylab="ERCC lib size/Total lib size",xlab="Proportion Zeroes in Single Cells")
abline(v=0.85,h=0.01,lty=2)

```

## Filtering out cells
We decide to filter out cells with > 85% dropout, smaller than 10000 library size, as cells that had a ratio of ERCC library size to total library size that had > 0.01.
```{r}
keep1 <- dropout<0.85 & lib.size.end>10000 & lib.size.ercc/lib.size < 0.01
counts.keep <- counts.end[,keep1]
dim(counts.keep)
```
This leaves `r ncol(counts.keep)` cells for further analysis.

## Filtering out genes
We chose to filter out genes that had more than 90% zeroes.
```{r}
# Filter out genes where > 90% of the cells are zeroes
propZ_genes <- rowSums(counts.keep==0)/ncol(counts.keep)
counts.keep <- counts.keep[propZ_genes<=0.9,]
dim(counts.keep)
```
This left `r nrow(counts.keep)` genes for further analysis. We finally check an MDSplot with PC1 and PC2.
```{r}
par(mfrow=c(1,1))
plotMDS(DGEList(counts.keep),pch=16,gene.selection = "common")
```

# Converting to DGEList, adding gene annotation
```{r,cache=TRUE}
y <- DGEList(counts.keep)
```

```{r,cache=TRUE}
#annotation
symbol <- toTable(org.Hs.egSYMBOL)
m <- match(rownames(y),symbol$symbol)
ann <- data.frame(Original_ID=rownames(y),symbol[m,])
rownames(ann) <- rownames(y)
ens <- toTable(org.Hs.egENSEMBL)
m <- match(ann$gene_id,ens$gene_id)
ann$ensembl_id <- ens$ensembl_id[m]
chr <- toTable(org.Hs.egCHR)
m <- match(ann$gene_id,chr$gene_id)
ann$chr <- chr$chromosome[m]
genename <- toTable(org.Hs.egGENENAME)
m <- match(ann$gene_id,genename$gene_id)
ann$genename <- genename$gene_name[m]
m <- match(ann$ensembl_id,hg38.length$EnsID)
ann$length <- hg38.length$Length[m]

y$genes <- ann
```

# Filtering out additional genes based on annotation
We decided to filter out mitochondrial and ribosomal genes, which tend to be very short, as well as genes that lack an Entrez gene id and gene length annotation. This was to clean up the dataset as much as possible and stick to well annotated genes.

```{r}
mito <- grep("mitochondrial",y$genes$genename)
ribo <- grep("ribosomal",y$genes$genename)
chrm <- grep("MT",y$genes$chr)
junk <- unique(c(mito,ribo,chrm))
length(junk)
y <- y[-junk,]
y <- y[!is.na(y$genes$gene_id),]
y <- y[!is.na(y$genes$length),]
dim(y)
y$samples$lib.size <- colSums(y$counts)
```
After removing these genes, there are `r nrow(y)` genes left for further analysis.

# Gene length bias

```{r,cache=TRUE}
lcpm <- normCounts(y,log=TRUE,prior.count=1)
lrpkm <- rpkm(y,log=TRUE,prior.count=1)
propZ_genes <- rowSums(y$counts==0)/ncol(y)
```

```{r}
# Binning according to gene length
q<-quantile(sqrt(y$genes$length),probs=seq(0.1,1,0.1))
decile <- rep(NA,nrow(y))
decile[sqrt(y$genes$length)<=q[1]] <- 1
for(i in 2:10) decile[sqrt(y$genes$length)>q[i-1] & sqrt(y$genes$length)<=q[i]] <- i
```

# Figure for paper:

```{r,fig1A,cache=FALSE,fig.width=12,fig.height=5}
par(mar=c(8.5,4.5,3,2))
par(mfrow=c(1,3))
par(mgp=c(3,1,0))

q2 <- c(0,q^2)
labels <- rep(NA,10)
for(i in 1:10) labels[i] <-paste(round(q2[i]),"-",round(q2[i+1]),sep="")

boxplot(rowMeans(lcpm)~decile,ylab="AvgLogCounts",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(0,4))
title("(g) Human K562,  droplet umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(propZ_genes~decile,ylab="Proportion zeroes in each gene",cex.lab=1.5,cex.axis=1.2,names=labels,las=2)
title("(h) Human K562,  droplet umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(rowMeans(lrpkm)~decile,ylab="AvgLogRPKM",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(0,12))
title("(i) Human K562,  droplet umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
```

```{r}
save(lcpm,decile,labels,propZ_genes,lrpkm,file="./figuresForPaper/Klein-fig2C.Rdata")
```


