---
title: "Kolodziejczyk-mESC-analysis"
author: "Belinda Phipson"
date: "06/02/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This mouse ESC dataset comes from Kolodziejczyk AA et al. 2015. Single cell RNA-sequencing of pluripotent states unlocks modular transcriptional variation. Cell Stem Cell. 17:471–485:  http://www.sciencedirect.com/science/article/pii/S193459091500418X.

Details of the dataset and processing directly from the paper:

##Cell Culture of mESCs

Cells were grown in three media:

The three media are as follows:


1. Serum-containing media: Knockout DMEM (GIBCO), 1X penicillin-streptomycin-glutamine (GIBCO), 1X non-essential amino acids (GIBCO), 100 U/ml recombinant human leukemia inhibitory factor (Millipore), 15% fetal bovine serum (HyClone), 0.1mM $\beta$-mercaptoethanol (Sigma).


2. Standard 2i media: N2B27 basal media (NDiff 227, StemCells), 100 U/ml recombinant human LIF (Millipore), 1 $\mu$ M PD0325901 (Stemgent), 3 $\mu$ M CHIR99021 (Stemgent).
  
3.  Alternative 2i media: N2B27 basal media (NDiff 227, StemCells), 100 U/ml recombinant human LIF (Millipore), 1 $\mu$ M CGP77675 (Sigma), 3 $\mu$ M CHIR99021 (Stemgent).

##cDNA Library Preparation from Single Cells using the Fluidigm C1

For each culture condition, 4,000 cells were loaded on to a 10–17 $\mu$ m Fluidigm C1 Single-Cell Auto Prep IFC, and cell capture was performed according to the manufacturer’s instructions. The capture efficiency was inspected using a microscope to remove samples from the analysis with more than one cell captured. Upon capture, reverse transcription and cDNA preamplification were performed in the 10–17 $\mu$ m Fluidigm C1 Single-Cell Auto Prep IFC using the SMARTer PCR cDNA Synthesis Kit (Clontech) and the Advantage 2 PCR Kit (Ramskold et al., 2012). cDNA was harvested and diluted to a range of 0.1–0.3 ng/$\mu$ l and Nextera libraries were prepared using the Nextera XT DNA Sample Preparation Kit and the Nextera Index Kit (Illumina) following the instructions in the Fluidigm manual “Using the C1™ Single-Cell Auto Prep System to Generate mRNA from Single Cells and Libraries for Sequencing.” Libraries from one chip were pooled, and paired-end 100 bp sequencing was performed on four lanes of an Illumina HiSeq2000.

## Bulk RNA-sequencing 
Libraries were prepared and sequenced using the Wellcome Trust Sanger Institute sample preparation pipeline with Illumina’s TruSeq RNA Sample Preparation v2 Kit. RNA was extracted from 1–2 million cells using the QIAGEN RNA Purification Kit on a QiaCube robot. The quality of the RNA sample was checked using gel electrophoresis. For library preparation, poly-A RNA was purified from total RNA using oligo-dT magnetic pull-down. Subsequently, mRNA was fragmented using metal-ion catalyzed hydrolysis. The cDNA was synthesized using random hexamer priming, and end repair was performed to obtain blunt ends. A-tailing was done to enable subsequent ligation of Illumina paired-end sequencing adapters, and samples were multiplexed at this stage. The resulting library was amplified using 10 cycles of PCR, substituting the Kapa Hifi polymerase for the polymerase in the Illumina TruSeq Kit. Samples were diluted to 4nM, and 100 bp paired-end sequencing was carried out on an Illumina HiSeq2000. Sequencing Quality Control was performed by the Sanger sequencing facility.

# Processing

The data was downloaded from the ArrayExpress database (http://www.ebi.ac.uk/arrayexpress), under accession number E-MTAB-2600 and processed by Luke Zappia using a bpipe pipeline. The quality of the raw sequencing reads was examined using FastQC v0.11.4 and they were checked for contamination by aligning a sample of reads to multiple reference genome using FastQ Screen v0.6.4. Reads were aligned to the mm10 mouse genome using STAR v2.5.2a and reads overlapping genes in an appropriate annotation (Gencode M9 for mouse) were counted using featureCounts v1.5.0-p3. This pipeline was constructed in Bpipe v0.9.9.3 and a report summarising the steps produced using MultiQC v0.8. 

# Load R libraries, and useful R functions

```{r,libraries,cache=FALSE}
library(edgeR)
library(RColorBrewer)
library(org.Mm.eg.db)
library(limma)
```
```{r,myRfunctions}
source("/group/bioi1/belinda/SingleCell/Normalisation/normCounts.R")
```
```{r}
ggplotColors <- function(g){

  d <- 360/g

  h <- cumsum(c(15, rep(d,g - 1)))

  hcl(h = h, c = 100, l = 65)

}
```

# Reading in the data and sample information

```{r,cache=TRUE}
counts.mesc <- read.delim("/group/bioi1/shared/public_data/Kolodziejczyk-mESCs-scRNAseq/counts/counts.txt",skip=1,stringsAsFactors = FALSE)
targets.mesc <- read.delim("/group/bioi1/shared/public_data/Kolodziejczyk-mESCs-scRNAseq/metadata/E-MTAB-2600.sdrf.txt",stringsAsFactors = FALSE)
d <- duplicated(targets.mesc$Comment.ENA_RUN.)
targets.mesc <- targets.mesc[!d,]

# Clean up column names
colnames(counts.mesc) <- gsub("X\\.mnt\\.storage\\.shared\\.public_data\\.Kolodziejczyk\\.mESCs\\.scRNAseq\\.STAR\\.2pass\\.","",colnames(counts.mesc))
colnames(counts.mesc) <- gsub("\\.Aligned\\.out\\.bam","",colnames(counts.mesc))

y.mesc <- counts.mesc[,-(1:6)]
m <- match(colnames(y.mesc),targets.mesc$Comment.ENA_RUN.)
targets.mesc <- targets.mesc[m,]
table(targets.mesc$Comment.ENA_RUN.==colnames(y.mesc))
rownames(y.mesc) <- counts.mesc[,1]
```

```{r,cache=TRUE}
# Separating endogenous and ERCC controls
ercc <- grep("ERCC",rownames(y.mesc))
y.mesc.ercc <- y.mesc[ercc,]
y.mesc.end <- y.mesc[-ercc,]
```

# Quality control of the cells
## QC plots
Before any analysis, the quality of the cells was examined by looking at metrics such as dropout and total library size of each cell. We also looked at the proportion of the sequencing taken up by the ERCC controls. There are 5 bulk controls to use for QC as well.

```{r}
# Calculate dropout and library size
dropout <- colSums(y.mesc.end==0)/nrow(y.mesc.end)
lib.size <- colSums(y.mesc)
lib.size.ercc <- colSums(y.mesc.ercc)
lib.size.end <- colSums(y.mesc.end)
```

```{r,cache=TRUE}
mypal <- brewer.pal(3,"Dark2")
par(mfrow=c(1,2))
plot(dropout,sqrt(lib.size.end),col=mypal[factor(targets.mesc$Material.Type)],xlab="Proportion Zeroes in Single Cells",ylab="Sqrt(library size)")
legend("topright",legend=c("Bulk","SingleCell"),fill=mypal[1:2])
plot(dropout,sqrt(lib.size.end),col=mypal[factor(targets.mesc$FactorValue.growth.condition.)],xlab="Proportion Zeroes in Single Cells",ylab="Sqrt(library size)")
title("Coloured by the 3 growth media")
abline(h=750,lty=2)
abline(v=0.8,lty=2)

plot(dropout,lib.size.ercc/lib.size,col=mypal[factor(targets.mesc$FactorValue.growth.condition.)],xlim=c(0.6,1),ylab="ERCC lib size/Total lib size",xlab="Proportion Zeroes in Single Cells")
legend("topleft",legend=levels(factor(targets.mesc$FactorValue.growth.condition.)),pch=16,col=mypal,cex=0.8,bty="n")
title("Coloured by 3 growth media")
abline(h=0.28^2,lty=2)
abline(v=0.8,lty=2)

# Colour by plate
mypal <- brewer.pal(8,"Dark2")
morecol <- colorRampPalette(mypal)
cols <- morecol(14)

plate <- gsub("_.$","",targets.mesc$Source.Name)
plate <- gsub("_..$","",plate)
table(plate)
targets.mesc$plate <- plate

plot(dropout,lib.size.ercc/lib.size,col=cols[factor(targets.mesc$plate)],xlim=c(0.6,1),ylab="ERCC lib size/Total lib size",xlab="Proportion Zeroes in Single Cells")
title("Coloured by C1 plate")
legend("topleft",legend=levels(factor(targets.mesc$plate)),pch=16,col=cols,bty="n",cex=0.8)
abline(h=0.28^2,lty=2)
abline(v=0.8,lty=2)

```

## Filtering out cells

Based on the plots, we decided to stringently filter out cells that have a very high proportion of zeroes and very low library size (Prop > 0.8 and library size < 500000). We also decided to filter out the three plates that had very high proportions of ERCC spike ins: 2i\_3, a2i\_3, serum\_3. After filtering, we are left with 530 single cells for further analysis.

```{r,cache=TRUE}
# Keep cells with prop zeroes less than 0.8 and library size > 500000
keep1 <- dropout<0.8 & lib.size.end>500000
y.mesc.keep <- y.mesc.end[,keep1]
targets.keep <- targets.mesc[keep1,]

# Get rid of those three bad plates
bad_plates <- c("2i_3","a2i_3","serum_3")
throw_away <- targets.keep$plate %in% bad_plates
y.mesc.keep <- y.mesc.keep[,!throw_away]
targets.keep <- targets.keep[!throw_away,] 

# Separate bulk from single cells
y.bulk <- y.mesc.keep[,targets.keep$Material.Type=="cells"]
y.mesc.keep <- y.mesc.keep[,targets.keep$Material.Type=="single cell"]
targets.bulk <- targets.keep[targets.keep$Material.Type=="cells",]
targets.keep <- targets.keep[targets.keep$Material.Type=="single cell",]

table(targets.keep$Comment.ENA_RUN. == colnames(y.mesc.keep))
```

## Filtering out genes
We chose to filter out genes that had more than 90% zeroes.
```{r}
# Filter out genes where > 90% of the cells are zeroes
propZ_genes <- rowSums(y.mesc.keep==0)/ncol(y.mesc.keep)
y.mesc.keep <- y.mesc.keep[propZ_genes<=0.9,]
dim(y.mesc.keep)
```

Finally we looked at plots of PC1 and PC2 to make sure our QC was adequate.
```{r}
# MDSplot
media <- targets.keep$FactorValue.growth.condition.
par(mfrow=c(1,2))
lnc <- normCounts(DGEList(y.mesc.keep),log=TRUE,prior.count=1)
plotMDS.default(lnc,pch=16,gene.selection = "common",col=ggplotColors(3)[factor(media)])
legend("bottom",legend=levels(factor(media)),pch=16,col=ggplotColors(3),cex=0.8)
title("coloured by media")

plotMDS.default(lnc,pch=16,gene.selection = "common",col=ggplotColors(6)[factor(targets.keep$plate)])
legend("bottom",legend=levels(factor(targets.keep$plate)),pch=16,col=ggplotColors(6),cex=0.8)
title("coloured by plate")
```

# Converting to DGEList, adding gene annotation
```{r,cache=TRUE}
y <- DGEList(y.mesc.keep,group=media)
rownames(y)<-gsub("\\..*","",rownames(y))
rownames(counts.mesc) <- gsub("\\..*","",counts.mesc[,1])
```

```{r,cache=TRUE}
#annotation
ens <- toTable(org.Mm.egENSEMBL)
m <- match(rownames(y),ens$ensembl_id)
ann <- data.frame(Original_ID=rownames(y),ens[m,])
rownames(ann) <- rownames(y)
symbol <- toTable(org.Mm.egSYMBOL)
m <- match(ann$gene_id,symbol$gene_id)
ann$symbol <- symbol$symbol[m]
chr <- toTable(org.Mm.egCHR)
m <- match(ann$gene_id,chr$gene_id)
ann$chr <- chr$chromosome[m]
genename <- toTable(org.Mm.egGENENAME)
m <- match(ann$gene_id,genename$gene_id)
ann$genename <- genename$gene_name[m]
m <- match(rownames(y),rownames(counts.mesc))
ann$length <- counts.mesc$Length[m]

y$genes <- ann
```

# Filtering out additional genes based on annotation
We decided to filter out mitochondrial and ribosomal genes, as well as genes that lack an Entrez gene id. This was to clean up the dataset as much as possible and stick to well annotated genes.

```{r}
mito <- grep("mitochondrial",y$genes$genename)
ribo <- grep("ribosomal",y$genes$genename)
chrm <- grep("MT",y$genes$chr)
junk <- unique(c(mito,ribo,chrm))
length(junk)
y <- y[-junk,]
y <- y[!is.na(y$genes$gene_id),]
dim(y)
y$samples$lib.size <- colSums(y$counts)
y.kol <- y
save(y.kol, file="./mESCs-data/kol.Rdata")
```

# Gene length bias
We binned genes into 10 roughly equal bins based on gene length, with roughly 1200 genes in each bin.
```{r,cache=TRUE}
lcpm <- normCounts(y,log=TRUE,prior.count=1)
lrpkm <- rpkm(y,log=TRUE,prior.count=1)
propZ_genes <- rowSums(y$counts==0)/ncol(y)
```

```{r}
# Binning according to gene length
q<-quantile(sqrt(y$genes$length),probs=seq(0.1,1,0.1))
decile <- rep(NA,nrow(y))
decile[sqrt(y$genes$length)<=q[1]] <- 1
for(i in 2:10) decile[sqrt(y$genes$length)>q[i-1] & sqrt(y$genes$length)<=q[i]] <- i
```

# Figure for paper:

```{r,fig1A,cache=FALSE,fig.width=12,fig.height=5}
par(mar=c(8.5,4.5,3,2))
par(mfrow=c(1,3))
par(mgp=c(3,1,0))

q2 <- c(0,q^2)
labels <- rep(NA,10)
for(i in 1:10) labels[i] <-paste(round(q2[i]),"-",round(q2[i+1]),sep="")

boxplot(rowMeans(lcpm)~decile,ylab="AvgLogCounts",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(0,8))
title("(a) Mouse ESCs, full-length",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(propZ_genes~decile,ylab="Proportion zeroes in each gene",cex.lab=1.5,cex.axis=1.2,names=labels,las=2)
title("(b) Mouse ESCs, full-length",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(rowMeans(lrpkm)~decile,ylab="AvgLogRPKM",cex.lab=1.5,cex.axis=1.2,names=labels,las=2)
title("(c) Mouse ESCs, full-length",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
```


