---
title: "Zeigenhain-mESCs-2016"
author: "Belinda Phipson"
date: "08/02/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This data is mouse embryonic stem cells from Ziegenhain, C. et al. Comparative analysis of single-cell RNA sequencing methods. bioRxiv (2016).

I downloaded the data from GEO under accession GSE75790: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75790

There are five different datasets, three of which inclue UMIs. The authors ran the UMI datasets through a dropseq pipeline, which includes mapping using STAR. For the purpose of the paper we are only analysing the SCRB-Seq UMI data, which has the largest number of cells.

All datasets were generated by running the cells in two batches, hence there is a batch effect in the data.

# Load R libraries, and useful R functions

```{r,libraries,cache=FALSE}
library(edgeR)
library(RColorBrewer)
library(org.Mm.eg.db)
library(limma)
```
```{r,myRfunctions}
source("/group/bioi1/belinda/SingleCell/Normalisation/normCounts.R")
```
```{r}
ggplotColors <- function(g){

  d <- 360/g

  h <- cumsum(c(15, rep(d,g - 1)))

  hcl(h = h, c = 100, l = 65)

}
```
# Reading in the data

```{r,cache=TRUE}
counts <- read.delim("/group/bioi1/shared/public_data/meSCS-scRNASeq-UMI-Grun2014/GSE75790_ziegenhain_complete_data.txt",stringsAsFactors = FALSE,header=TRUE)
prot <- gsub("_.*","",colnames(counts))
scrb <- counts[,prot=="SCRBseq"]

# load mm10 gene length information
mm10.length <- read.table("mm10-length.txt",header=TRUE,stringsAsFactors = FALSE)
```

```{r}
# Separate ercc and endogenous genes
ercc <- grep("ERCC", rownames(scrb))
counts.end <- scrb[-ercc,]
counts.ercc <- scrb[ercc,]
```

# Quality control of the cells
## QC plots

Before any analysis, the quality of the cells was examined by looking at metrics such as dropout and total library size of each cell. We also looked at the proportion of the sequencing taken up by the ERCC controls. 

```{r}
# Calculate dropout and library size
dropout <- colSums(counts.end==0)/nrow(counts.end)
lib.size <- colSums(scrb)
lib.size.ercc <- colSums(counts.ercc)
lib.size.end <- colSums(counts.end)
```

```{r}
par(mar=c(5,4,2,2))
par(mfrow=c(1,2))
plot(dropout,lib.size.end,xlab="Proportion Zeroes in Single Cells",ylab="Library size")
plot(dropout,lib.size.ercc/lib.size,ylab="ERCC lib size/Total lib size",xlab="Proportion Zeroes in Single Cells")
```

## Filtering of the cells
In this dataset, there aren't too many obvious outlier cells when looking at the QC plots. Perhaps the cells were already filtered before the data was submitted to GEO. I decided to keep all cells.

## Filtering out genes
We chose to filter out genes that had more than 90% zeroes.
```{r}
# Filter out genes where > 90% of the cells are zeroes
propZ_genes <- rowSums(counts.end==0)/ncol(counts.end)
counts.keep <- counts.end[propZ_genes<=0.9,]
dim(counts.keep)
```
This left `r nrow(counts.keep)` genes for further analysis. We finally check an MDSplot with PC1 and PC2.

```{r}
batch <- rep(NA,ncol(counts.keep))
batch[grep("SCRBseqA",colnames(counts.keep))] <- "batchA"
batch[grep("SCRBseqB",colnames(counts.keep))] <- "batchB"
plotMDS(DGEList(counts.keep),pch=16,gene.selection = "common",col=ggplotColors(2)[factor(batch)])
legend("topleft",legend=levels(factor(batch)),fill=ggplotColors(2))
```


# Converting to DGEList, adding gene annotation
```{r,cache=TRUE}
y <- DGEList(counts.keep,group=batch)
```
```{r,cache=TRUE}
#annotation
ens <- toTable(org.Mm.egENSEMBL)
m <- match(rownames(y),ens$ensembl_id)
ann <- data.frame(Original_ID=rownames(y),ens[m,])
rownames(ann) <- rownames(y)
symbol <- toTable(org.Mm.egSYMBOL)
m <- match(ann$gene_id,symbol$gene_id)
ann$symbol <- symbol$symbol[m]
chr <- toTable(org.Mm.egCHR)
m <- match(ann$gene_id,chr$gene_id)
ann$chr <- chr$chromosome[m]
genename <- toTable(org.Mm.egGENENAME)
m <- match(ann$gene_id,genename$gene_id)
ann$genename <- genename$gene_name[m]
m <- match(rownames(y),mm10.length$EnsID)
ann$length <- mm10.length$Length[m]

y$genes <- ann
```

# Filtering out additional genes based on annotation
We decided to filter out mitochondrial and ribosomal genes, as well as genes that lack an Entrez gene id. This was to clean up the dataset as much as possible and stick to well annotated genes.

```{r}
mito <- grep("mitochondrial",y$genes$genename)
ribo <- grep("ribosomal",y$genes$genename)
chrm <- grep("MT",y$genes$chr)
junk <- unique(c(mito,ribo,chrm))
length(junk)
y <- y[-junk,]
y <- y[!is.na(y$genes$gene_id),]
y <- y[!is.na(y$genes$length),]
dim(y)
y$samples$lib.size <- colSums(y$counts)
y.zeig <- y
save(y.zeig,file="./mESCs-data/zeig.Rdata")
```
This leaves `r ncol(y)` cells and `r nrow(y)` genes for further analysis.

# Gene length bias
We binned genes into 10 roughly equal bins based on gene length, with roughly 1000 genes in each bin.
```{r,cache=TRUE}
lcpm <- normCounts(y,log=TRUE,prior.count=1)
lrpkm <- rpkm(y,log=TRUE,prior.count=1)
propZ_genes <- rowSums(y$counts==0)/ncol(y)
```

```{r}
# Binning according to gene length
q<-quantile(sqrt(y$genes$length),probs=seq(0.1,1,0.1))
decile <- rep(NA,nrow(y))
decile[sqrt(y$genes$length)<=q[1]] <- 1
for(i in 2:10) decile[sqrt(y$genes$length)>q[i-1] & sqrt(y$genes$length)<=q[i]] <- i
```

# Figure for paper:

```{r,fig1A,cache=FALSE,fig.width=12,fig.height=5}
par(mar=c(8.5,4.5,3,2))
par(mfrow=c(1,3))
par(mgp=c(3,1,0))

q2 <- c(0,q^2)
labels <- rep(NA,10)
for(i in 1:10) labels[i] <-paste(round(q2[i]),"-",round(q2[i+1]),sep="")

boxplot(rowMeans(lcpm)~decile,ylab="AvgLogCounts",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(0,5))
title("(a) Mouse ESCs, umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(propZ_genes~decile,ylab="Proportion zeroes in each gene",cex.lab=1.5,cex.axis=1.2,names=labels,las=2)
title("(b) Mouse ESCs, umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
boxplot(rowMeans(lrpkm)~decile,ylab="AvgLogRPKM",cex.lab=1.5,cex.axis=1.2,names=labels,las=2,ylim=c(-1,10))
title("(c) Mouse ESCs, umi",cex.main=1.5)
title(xlab="Gene length",cex.lab=1.5,line=7)
```
```{r}
save(lcpm,decile,labels,propZ_genes,lrpkm,file="./figuresForPaper/Zeig-fig2Aii.Rdata")
```

